### DISCUSSION TITLE
Graphical representation of daily/monthly power?

### DISCUSSION THREAD

**Discussion by Elmi77** [arendst/Tasmota#16818] [Ideas]
I'm using a Tasmota-based switching socket for checking what power my solar balcony power plant produces. For a more exact analysis it would be nice to see what power is produced during what time of the day and during what months of the year. So what I'm requesting is a simple graph, which shows the total power generation over the day and over the year.

I know, I can use some external MQTT-services to implement this function, But beside the potential security problems, this is a logical problem: The idea is, to produce some green energy with the balcony solar power plant. Now using external services somewhere in the internet itself wastes a lot of energy. So there are situations, where the power produced by the solar power plant is smaller than what is consumed by the data transfer to and from such an external MQTT service. Means in such situations the pure existance of the solar power plant would consume power!

So I really would prefer to see such a function integrated in the local device only, without any power-consuming dependencies to external services.

**Comment by sfromis** [comment]
The basic setup for doing it via MQTT is certainly not going for some "external service", but doing it in-house, generally on the same machine already used for home automation, which can be a quite low-power server.

And having on-device historical data collection is not well-suited to most energy monitoring devices suitable for Tasmota.

**Comment by Elmi77** [comment]
There is also no in-house machine running. I have no home-automation crap, I'm just measuring what the solar power plant produces. Btw: having a separate computer running for gathering these data has exactly the same problem: additional power consumption where power should be produced.

**Reply by barbudor** [reply]
With Scripting Language you can have graphs on the front page but the esp8266 in your plug just have not enough memory to keep data across months.
You may upgrade for an ESP32 based device such as sonoff dual r2 or Shelly that generally comes with 4mb flash so enough to store data on file system
You would be able to have the graph, store some data for time and download back those data to a computer that is not always on for long term archives

**Reply by pkkrusty** [reply]
A raspberry pi Zero 2W consumes about 5w and can power all sorts of data collection. If you're dealing with such small production that you can't afford 5w and $15, you aren't going to find much capability for less.

**Reply by Elmi77** [reply]
> the esp8266 in your plug just have not enough memory to keep data across months.

The feature I'm requesting needs a total of 144 bytes: 12 int's for the total monthly statistics and 24 int's for the hourly statistics.

And to make that clear: I am definitely not interested in running an additional device, no matter how less it may consume. The switching Wifi plug has all the resources needed, so it makes absolutely no sense to have an additional power consumption thingy running. That's why I'm requesting a feature that is done within Tasmota software and not a fancy, super-duper, external power-eating solution.

**Reply by barbudor** [reply]
There is no need for additional feature
As we already said you already have Scripting-Language to draw graphs on the web page

Did you search the doscs and searched about our suggestions ?
https://tasmota.github.io/docs/Scripting-Language/
https://github.com/arendst/Tasmota/discussions/15063
https://github.com/arendst/Tasmota/discussions/15767

However, would for any reason your Tasmota restarts, you will loose everything.
You have been warned

**Reply by Elmi77** [reply]
> There is no need for additional feature

Are you the developer of Tasmota? Or based on what do you decide this?

> However, would for any reason your Tasmota restarts, you will loose everything.

Sorry, but that is complete nonsense - the yesterdays and total energy counter also does not get lost on restart.

**Reply by pkkrusty** [reply]
There are multiple developers, but as Tasmota is open source, anyone can submit PRs to improve functionality. If someone deems your suggestion worthy of time, they may code it, but it's a pretty specific use-case. The main developers are pretty clear that flash space is tightly controlled. Unless you code it yourself, you hope that someone else finds your suggestion worthy.

Barbudor is saying that something pretty close to what you want already exists, but you will have to take the time to adapt the code to your specific situation. Tasmota has the ability with scripting to show graphs and totals on the WebUI. 

With scripting, things are stored in RAM unless you store in one of the 16 mem variables. You could do some trickery to store all the data you need in the mem variables and read it out again as needed. 

> Sorry, but that is complete nonsense - the yesterdays and total energy counter also does not get lost on restart.

The ESP8266 will lose whatever is in RAM on a restart. The power monitoring device may have registers that store energy data persistently that the ESP device could re-access, but is it enough to re-build your monthly graph? Or even hourly?

**Reply by Elmi77** [reply]
> There are multiple developers, but as Tasmota is open source, anyone can submit PRs to improve functionality. If someone deems your suggestion worthy of time, they may code it, but it's a pretty specific use-case. The main developers are pretty clear that flash space is tightly controlled. Unless you code it yourself, you hope that someone else finds your suggestion worthy.

Sorry, but please explain this: when I try to file a feature request on Github, I'm forwarded to this forum. So following the documentation on Github, I filed my feature request here - and now I'm blamed for doing this? So what is the exact procedure to file a feature request that is really read by the DEVELOPERS but not by some random guys?

> The power monitoring device may have registers that store energy data persistently that the ESP device could re-access, but is it enough to re-build your monthly graph? Or even hourly?

I already explained this, I'm talking about 144 bytes. I'm not familar with the data structures within Tasmota, but It would be really nice when a **developer** could confirm if this space is available in Flash or not.

**Reply by barbudor** [reply]
You are asking for something which is already there
Reading the docs that I pointed

144 bytes is too much for Tasmota, we don't have that available. Any PR that will request 144 Bytes would be rejected

**Reply by Elmi77** [reply]
> You are asking for something which is already there Reading the docs that I pointed

It is NOT there. Wasting additional 120 Wh per day / 43,8 kWh per year while there is already an device available which could do that job is simply weird.
 
> 144 bytes is too much for Tasmota, we don't have that available. Any PR that will request 144 Bytes would be rejected

Once again: you are what or who to decide this? Is your name Theo Arends?

**Reply by pkkrusty** [reply]
Seems like you're getting a little frustrated. You're welcome to send a message to Theo and see what he says, but for Barbudor, who has been coding alongside him for years, to say that 144 bytes is too much, I tend to believe him. 

You say the solution is not there, users with much more experience than you are saying it is there. Who to believe?

**Reply by sfromis** [reply]
Reserving 144 settings bytes for a very narrow feature is not going to happen with the Tasmota project. Obviously, with open source, you are free to have a private fork where you can have it your way.

But still.... One "trick" to have more than the 16 permanent variables is to reuse strings like FriendlyName to house other things, meaning that you can get 144 bytes indirectly, if you are a bit careful with other string pool usage.

**Reply by barbudor** [reply]
The 16 Mem are part of Rules, not available with Scripting-Language which is required to have charts on the WebGUI
Scripting has support for permament vars by re-using the room but that's limited to 50 bytes
The simplest way to go over would be to use file system but that would require a 4MB flash which is not present on cheap power plug.
Would be the case on an ESP32 based device.
Of course some soldering skills could allow to replace device flash with a 4MB, given it's not a ESP8285 of course.

But of course, that's just a comment from a random guy 🤣

**Reply by pkkrusty** [reply]
@sfromis i like that trick. I'll have to try it when I have some downtime next week.

**Comment by arendst** [comment]
@Elmi77 I just chime in to assure you that the responses you received here are totally supported by me.

Pls try the suggested script solution or provide a PR demonstrating the minimal code size increase of just 144 bytes....
